src/bin2d.js

```diff
@@ -0,0 +1,145 @@
+import {slice} from "./array";
+import bisect from "./bisect";
+import constant from "./constant";
+import extent from "./extent";
+import range from "./range";
+import {tickStep} from "./ticks";
+import sturges from "./threshold/sturges";
+
+export default function() {
```

Comment:
I'll leave it to @mbostock to comment on coding style.

My thoughts here are - have you tried implementing `bin2d` by composing results from `bin`? Probably `bin2D` could be implemented by leveraging the existing 1D binning features, rather than including an entire separate binning implementation with lots of overlapping logic.

Reply:
Composing results from bin would require to scan the entire cross of all X by all Y bins. This implementation is closer to hexbin and more efficient although not optimized with the latest ES features.

Reply:
@curran, @Fil 's experiment goes into the direction you pointed to, adding x0 and x1 iteratively to each data element for a dimension when calling d3.bin. The complexity comes when grouping back all the elements per unique combination of each dimension's x0,x1 properties. 
The implementation I provided is fairly mechanical, but works. Overall, the implementation is 146 lines with 88 lines for bin2d including comments.